### ISSUE TITLE
[Question] Discord to Matrix bridge

### ISSUE THREAD

**Issue by TitanNano** [godot-rust/gdext#313]
Sorry for abusing issues for this, but unfortunately, discussions are not enabled on this repo.

I'm aware that most discussions are happening on discord, but unfortunately this is extremely in accessible for outsiders and especially for non-discord-contributors. 

Could whoever is running the discord server consider adding a matrix bridge to the discord server so that everyone can take part in the conversations?

**Comment by TitanNano** [comment]
I have created a matrix room (private for now) and can set up a bridge via [t2bot.io](https://t2bot.io) but it has be approved by a discord admin, so the not can join that side of the bridge. I will also have to repeat this for each room that should be bridged.

**Comment by Bromeon** [comment]
Thanks for bringing this up. 
I hope you don't mind if I ask some concrete questions 🙂 
1. What kind of permissions would such a bridge need? (Also from a security perspective)
2. Is everything synced 1:1 or is it possible to have extra channels in Matrix that are not part of Discord?
3. How are messages persisted, and for how long? What happens if messages are edited/deleted?
4. What about moderation, kicks, bans etc?
5. Discord offers some built-in spam and bot protection. Does Matrix have something similar? 

Please don't just refer to a link 😉  

Generally, there are always going to be some platforms that are inaccessible to others; even GitHub requires an account to participate. What I want to avoid is that the maintainer team has to spend time on moderating more platforms (even a synced bridge likely needs configuration, security settings, etc). So it also depends a bit how many users would actually benefit from this.

**Comment by lilizoey** [comment]
i do think it'd be good regardless to have what happens on discord be available to people who dont use discord though.

**Comment by TitanNano** [comment]
> 1. What kind of permissions would such a bridge need? (Also from a security perspective)

The brige bot requires the ` Manage Webhooks` role for full integration

> You can use Webhooks to make messages relayed by the bridge not nested by the bot user. This will also display the avatar of the user speaking on matrix with their messages.
> 
> The bot should create this automatically, but if not perform the following:
> - Enable Manage Webhooks on the role added by the bot.
> - Add the _matrix Webhook for each channel you'd like to enable this feature on.

-------------
> 2. Is everything synced 1:1 or is it possible to have extra channels in Matrix that are not part of Discord?

Bridging happens on a per room / channel basis. The bridge bot has to be added on both sides for each channel.
Matrix does not have the same concept of servers like Discord does. In matrix all channels have a global address (greatly simplified) and channels can be associated to spaces. The idea would be that we create a `godot-rust` space which in turn will be added to the existing `Godot` space. 

------------
> 3. How are messages persisted, and for how long? What happens if messages are edited/deleted?

Matrix is a federated messaging protocol with multiple home servers participating in the network (like email or mastodon). Message retention has therefore to be assumed to be unlimited. Individual home servers can purge old message histories, but as there can be any number of participating home servers, there is no guarantee that messages will ever go away.

Looking at the [bridge source code](https://github.com/t2bot/matrix-appservice-discord/tree/t2bot.io), it appears that all update and delete message events are being synchronized both ways.

----------
> 4. What about moderation, kicks, bans etc?

Can not be synchronized and should be performed on the native platform of the targeted user.

----------

> 5. Discord offers some built-in spam and bot protection. Does Matrix have something similar?

There is a very popular and official auto moderation bot implementation for matrix, and I found at least two publicly available instances that could be added to a room. I would go with the Mozilla auto moderator, as I'm on the mozilla.org home server myself. It should be noted though that this bot would enforce the [Mozilla Community Participation Guidelines](https://www.mozilla.org/en-US/about/governance/policies/participation/), where I don't know if that is acceptable for this project.

**Comment by TitanNano** [comment]
@Bromeon how would you like to proceed with this?

**Comment by Bromeon** [comment]
Before committing to that, I'd like to have some sort of a trial. Maybe some weeks with only private access to see if/how it works, whether message edits/deletions/moderations are propagated correctly, etc. Do you think that would be possible?

**Comment by TitanNano** [comment]
Sorry for the very late response. We can certainly do a private trial, the room on the matrix side will only be accessible with an invitation. I would suggest we also just bridge one room at first and see how it works out. 

Unfortunately, I was too optimistic when reading documentation on the moderation bots. It appears that none of the "public" bots are available without being added to their whitelist, which requires a room to be related to the specific project, there are no general purpose bots.

**Comment by Bromeon** [comment]
This deserves an update -- sorry for letting it slide for so long.

After more thoughts, I'd like to put this on hold. The current system with different platforms works well. We are represented on GitHub for all code-related issues and help tickets, on Discord for casual discussion and help requests, on Twitter and Mastodon for sporadic updates. Especially the former two take a lot of time to manage, moderate, configure properly and constantly tweak in order to provide the best user experience. For example, we are just now experimenting with Discord's "forum" feature for more structured help requests.

Linking other community platforms brings quite a few unknowns and risk. For example, users recently raised a valid point that many people post in Discord under the assumption that their messages stay there and are not republished elsewhere. While we can't exactly call that "privacy", it might still not be in the interest of everyone to forward the data.

Another related point that came up was that Discord discussions should be publicly archived somewhere, so they are indexable. Several people disagreed; there is no inherent reason to preserve every single conversation for eternity. A lot of Discord content revolves around help requests -- suggestions given there may not be best practices, or may quickly become outdated at the current pace. Old posts are never updated -- so if they are continuously crawled by search and AI engines, this might hurt the project by promoting bad/outdated suggestions. Instead, we should focus on extracting the important bits from conversations and properly rewrite them here. I also plan to bring the official documentation (book) to a much higher standard. Just today I added [several pages about registration](https://www.github.com/godot-rust/book/pull/26).

Last year, I invested a huge amount of time into "infra" work -- basically everything that is not code, but is required around the project to keep workflows efficient and unbureaucratic. A closer insight is available on the blog post [2023 in Review](https://godot-rust.github.io/dev/godot-rust-2023-review/), under section _Tooling and infrastructure_ -- and the community platforms are not even listed.

In short, after all that work, we have now a setup that works reasonably well. My "infra budget" is not infinite, and I would prefer spending my time in 2024 on things that concretely improves user experience, including:
- Library bugfixes and features
- Documentation
- Dogfooding (a simple game) to have first-hand experience
- Tooling that benefits the library (e.g. CI)
- Website improvements

I hope this is understandable. Maybe we can revisit such a proposal in the future, but realistically not this year.
Nevertheless, thanks for all the input!
